src/Modules/OrchardCore.Commerce/Drivers/AddressFieldDisplayDriver.cs

```diff
@@ -40,14 +60,39 @@ public class AddressFieldDisplayDriver : ContentFieldDisplayDriver<AddressField>
             GetEditorShapeType(context),
             viewModel =>
             {
+                viewModel.UserAddressToSave = field.UserAddressToSave;
+
                 viewModel.Regions = Regions.All;
                 viewModel.Provinces.AddRange(Regions.Provinces);
                 PopulateViewModel(field, viewModel, context.PartFieldDefinition);
             });
 
     public override async Task<IDisplayResult> UpdateAsync(AddressField field, IUpdateModel updater, UpdateFieldEditorContext context)
     {
-        await updater.TryUpdateModelAsync(field, Prefix, addressField => addressField.Address);
+        var viewModel = new AddressFieldViewModel();
+
+        if (await updater.TryUpdateModelAsync(viewModel, Prefix))
+        {
+            field.Address = viewModel.Address;
+
+            if (viewModel.ToBeSaved &&
+                !string.IsNullOrEmpty(viewModel.UserAddressToSave) &&
+                _hca.HttpContext?.User.Identity?.IsAuthenticated == true &&
+                await _userManager.GetUserAsync(_hca.HttpContext.User) is User user &&
+                user.Properties.TryGetValue(UserAddresses, out var userAddressesToken) &&
+                userAddressesToken is JObject userAddresses &&
+                userAddresses.TryGetValue(nameof(UserAddressesPart), out var partToken) &&
+                partToken is JObject part)
+            {
+                if (part[viewModel.UserAddressToSave] is not JObject)
+                {
+                    part[viewModel.UserAddressToSave] = JObject.FromObject(new AddressField());
+                }
+
+                ((JObject)part[viewModel.UserAddressToSave])![nameof(AddressField.Address)] = JToken.FromObject(viewModel.Address);
```

Comment:
It's necessary because `JObject[key]` returns `JToken`. But from the structure we know that this will be always `JObject` since it's a serialized `AddressPart`. If it isn't, then the data is corrupted and is should throw an exception.

Reply:
Okay then!